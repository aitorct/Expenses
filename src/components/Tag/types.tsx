import {StyleProp, TextStyle, ViewStyle} from 'react-native';

interface BaseTag {
  textStyle?: StyleProp<TextStyle>;
  style?: StyleProp<ViewStyle>;
}

interface IconTextTag extends BaseTag {
  icon: string;
  text: string | number;
}

interface IconTag extends BaseTag {
  icon: string;
  text?: never;
}

interface TextTag extends BaseTag {
  text: string | number;
  icon?: never;
}

export type TagType = IconTextTag | IconTag | TextTag;
