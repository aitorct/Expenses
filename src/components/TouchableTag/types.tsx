import {FlexStyle, StyleProp, TextStyle} from 'react-native';

export interface Props {
  text: string;
  onPress: () => void;
  active: boolean;
  textStyle: StyleProp<TextStyle>;
  buttonStyle: StyleProp<FlexStyle>;
}
