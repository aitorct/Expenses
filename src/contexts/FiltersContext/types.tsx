export interface FiltersContextInterface {
  filters: Filters;
  applyFilters: (key: FilterType, value: FilterValue) => void;
  filtersActive: boolean;
}

export enum FilterType {
  USER = 'user',
  YEAR = 'year',
  RECEIPTS = 'receipts',
}

export interface Filter {
  value: FilterValue;
}

export type FilterValue = string | null;

export type Filters = Record<FilterType, Filter>;
