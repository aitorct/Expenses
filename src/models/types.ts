export interface ObjectExpenseData {
  id: string;
  amount_value: string;
  amount_currency: string;
  date: string;
  date_and_time: string;
  year: number;
  merchant: string;
  receipts: string[];
  comment: string;
  category: string;
  user_name: string;
  user_email: string;
}

export interface ExpenseData {
  id: string;
  amount: {
    value: string;
    currency: string;
  };
  date: string;
  merchant: string;
  receipts: Receipt[];
  comment: string;
  category: string;
  user: {
    first: string;
    last: string;
    email: string;
  };
}

export interface Receipt {
  url: string;
}

export interface ParsedExpense {
  id: string;
  amount: string;
  time: string;
  date: string;
  date_and_time: string;
  year: number;
  merchant: string;
  hasAttachedReceipts: boolean;
  receiptsAmount: number;
  receipts: string[];
  hasComment: boolean;
  comment: string;
  category: string;
  user_name: string;
  user_email: string;
}
