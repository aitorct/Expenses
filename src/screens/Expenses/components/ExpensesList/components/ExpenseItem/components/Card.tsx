import React from 'react';
import {StyleSheet, Text, View} from 'react-native';
import Tag from '@components/Tag';
import {ParsedExpense} from '@models/types';
import colors from '@theme/colors';
import fontSizes from '@theme/fontSizes';
import sizes from '@theme/sizes';

interface Props {
  expense: ParsedExpense;
}

const Card = ({expense}: Props) => {
  return (
    <View>
      <View style={styles.spaceBetweenRow}>
        <View>
          <Text style={styles.merchant}>{expense.merchant}</Text>
          <Text style={styles.time}>{expense.time}</Text>
        </View>
        <Text style={styles.amount}>{expense.amount}</Text>
      </View>
      <View style={[styles.spaceBetweenRow, styles.separator]}>
        <Tag icon={'account-circle'} text={expense.user_name} />
        <View style={styles.row}>
          {expense.hasComment && <Tag icon={'comment'} />}
          {expense.hasAttachedReceipts && (
            <Tag
              style={styles.tagPadding}
              icon={'receipt'}
              text={expense.receiptsAmount}
            />
          )}
        </View>
      </View>
    </View>
  );
};

const styles = StyleSheet.create({
  spaceBetweenRow: {
    flexDirection: 'row',
    justifyContent: 'space-between',
  },
  row: {
    flexDirection: 'row',
  },
  separator: {
    marginTop: sizes.s16,
  },
  merchant: {
    color: colors.text,
    fontWeight: 'bold',
    fontSize: fontSizes.text,
  },
  time: {
    color: colors.text,
    fontSize: fontSizes.smallText,
  },
  amount: {
    color: colors.text,
    fontWeight: '500',
    fontSize: fontSizes.text,
  },
  tagPadding: {
    marginLeft: sizes.s8,
  },
});

export default Card;
