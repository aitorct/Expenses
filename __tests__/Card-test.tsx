import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';
import {ParsedExpense} from '@models/types';
import Card from '@screens/Expenses/components/ExpensesList/components/ExpenseItem/components/Card';

const mockedExpense: ParsedExpense = {
  id: '1',
  amount: '18â‚¬',
  time: '17:48',
  date: '01/01/2020',
  date_and_time: '01/01/2020 17:48',
  year: 2018,
  merchant: 'Amazon',
  hasAttachedReceipts: false,
  receiptsAmount: 0,
  receipts: [],
  hasComment: false,
  comment: '',
  category: '',
  user_name: 'Aitor Cubeles Torres',
  user_email: 'aitor.cubeles@utexas.edu',
};

it('Card with no receipts and no comments renders correctly', () => {
  const tree = renderer.create(<Card expense={mockedExpense} />).toJSON();
  expect(tree).toMatchSnapshot();
});

it('Card with comments renders correctly', () => {
  const mockedExpenseWithComments = {
    ...mockedExpense,
    comment: 'This is a comment',
    hasComment: true,
  };

  const tree = renderer
    .create(<Card expense={mockedExpenseWithComments} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('Card with receipts attached renders correctly', () => {
  const mockedExpenseWithReceipts = {
    ...mockedExpense,
    receipts: ['fakeUrl'],
    hasAttachedReceipts: true,
  };

  const tree = renderer
    .create(<Card expense={mockedExpenseWithReceipts} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('Card with receipts and comments attached renders correctly', () => {
  const mockedExpenseWithReceipts = {
    ...mockedExpense,
    comment: 'This is a comment',
    hasComment: true,
    receipts: ['fakeUrl'],
    hasAttachedReceipts: true,
  };

  const tree = renderer
    .create(<Card expense={mockedExpenseWithReceipts} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});
