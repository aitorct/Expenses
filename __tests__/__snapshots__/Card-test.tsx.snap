// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card with comments renders correctly 1`] = `
<View>
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <View>
      <Text
        style={
          Object {
            "color": "#000000",
            "fontSize": 16,
            "fontWeight": "bold",
          }
        }
      >
        Amazon
      </Text>
      <Text
        style={
          Object {
            "color": "#000000",
            "fontSize": 14,
          }
        }
      >
        17:48
      </Text>
    </View>
    <Text
      style={
        Object {
          "color": "#000000",
          "fontSize": 16,
          "fontWeight": "500",
        }
      }
    >
      18€
    </Text>
  </View>
  <View
    style={
      Array [
        Object {
          "flexDirection": "row",
          "justifyContent": "space-between",
        },
        Object {
          "marginTop": 16,
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#E7EFFA",
            "borderRadius": 36,
            "flexDirection": "row",
            "padding": 4,
            "paddingHorizontal": 8,
          },
          undefined,
        ]
      }
    >
      <Text
        allowFontScaling={false}
        selectable={false}
        style={
          Array [
            Object {
              "color": "#000000",
              "fontSize": 16,
            },
            undefined,
            Object {
              "fontFamily": "Material Design Icons",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            Object {},
          ]
        }
      >
        󰀉
      </Text>
      <View
        style={
          Object {
            "paddingLeft": 4,
          }
        }
      />
      <View
        style={
          Object {
            "flexDirection": "row",
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#000000",
              },
              undefined,
            ]
          }
        >
          Aitor Cubeles Torres
        </Text>
      </View>
    </View>
    <View
      style={
        Object {
          "flexDirection": "row",
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "backgroundColor": "#E7EFFA",
              "borderRadius": 36,
              "flexDirection": "row",
              "padding": 4,
              "paddingHorizontal": 8,
            },
            undefined,
          ]
        }
      >
        <Text
          allowFontScaling={false}
          selectable={false}
          style={
            Array [
              Object {
                "color": "#000000",
                "fontSize": 16,
              },
              undefined,
              Object {
                "fontFamily": "Material Design Icons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
        >
          󰅺
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`Card with no receipts and no comments renders correctly 1`] = `
<View>
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <View>
      <Text
        style={
          Object {
            "color": "#000000",
            "fontSize": 16,
            "fontWeight": "bold",
          }
        }
      >
        Amazon
      </Text>
      <Text
        style={
          Object {
            "color": "#000000",
            "fontSize": 14,
          }
        }
      >
        17:48
      </Text>
    </View>
    <Text
      style={
        Object {
          "color": "#000000",
          "fontSize": 16,
          "fontWeight": "500",
        }
      }
    >
      18€
    </Text>
  </View>
  <View
    style={
      Array [
        Object {
          "flexDirection": "row",
          "justifyContent": "space-between",
        },
        Object {
          "marginTop": 16,
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#E7EFFA",
            "borderRadius": 36,
            "flexDirection": "row",
            "padding": 4,
            "paddingHorizontal": 8,
          },
          undefined,
        ]
      }
    >
      <Text
        allowFontScaling={false}
        selectable={false}
        style={
          Array [
            Object {
              "color": "#000000",
              "fontSize": 16,
            },
            undefined,
            Object {
              "fontFamily": "Material Design Icons",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            Object {},
          ]
        }
      >
        󰀉
      </Text>
      <View
        style={
          Object {
            "paddingLeft": 4,
          }
        }
      />
      <View
        style={
          Object {
            "flexDirection": "row",
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#000000",
              },
              undefined,
            ]
          }
        >
          Aitor Cubeles Torres
        </Text>
      </View>
    </View>
    <View
      style={
        Object {
          "flexDirection": "row",
        }
      }
    />
  </View>
</View>
`;

exports[`Card with receipts and comments attached renders correctly 1`] = `
<View>
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <View>
      <Text
        style={
          Object {
            "color": "#000000",
            "fontSize": 16,
            "fontWeight": "bold",
          }
        }
      >
        Amazon
      </Text>
      <Text
        style={
          Object {
            "color": "#000000",
            "fontSize": 14,
          }
        }
      >
        17:48
      </Text>
    </View>
    <Text
      style={
        Object {
          "color": "#000000",
          "fontSize": 16,
          "fontWeight": "500",
        }
      }
    >
      18€
    </Text>
  </View>
  <View
    style={
      Array [
        Object {
          "flexDirection": "row",
          "justifyContent": "space-between",
        },
        Object {
          "marginTop": 16,
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#E7EFFA",
            "borderRadius": 36,
            "flexDirection": "row",
            "padding": 4,
            "paddingHorizontal": 8,
          },
          undefined,
        ]
      }
    >
      <Text
        allowFontScaling={false}
        selectable={false}
        style={
          Array [
            Object {
              "color": "#000000",
              "fontSize": 16,
            },
            undefined,
            Object {
              "fontFamily": "Material Design Icons",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            Object {},
          ]
        }
      >
        󰀉
      </Text>
      <View
        style={
          Object {
            "paddingLeft": 4,
          }
        }
      />
      <View
        style={
          Object {
            "flexDirection": "row",
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#000000",
              },
              undefined,
            ]
          }
        >
          Aitor Cubeles Torres
        </Text>
      </View>
    </View>
    <View
      style={
        Object {
          "flexDirection": "row",
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "backgroundColor": "#E7EFFA",
              "borderRadius": 36,
              "flexDirection": "row",
              "padding": 4,
              "paddingHorizontal": 8,
            },
            undefined,
          ]
        }
      >
        <Text
          allowFontScaling={false}
          selectable={false}
          style={
            Array [
              Object {
                "color": "#000000",
                "fontSize": 16,
              },
              undefined,
              Object {
                "fontFamily": "Material Design Icons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
        >
          󰅺
        </Text>
      </View>
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "backgroundColor": "#E7EFFA",
              "borderRadius": 36,
              "flexDirection": "row",
              "padding": 4,
              "paddingHorizontal": 8,
            },
            Object {
              "marginLeft": 8,
            },
          ]
        }
      >
        <Text
          allowFontScaling={false}
          selectable={false}
          style={
            Array [
              Object {
                "color": "#000000",
                "fontSize": 16,
              },
              undefined,
              Object {
                "fontFamily": "Material Design Icons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
        >
          󰑉
        </Text>
        0
        0
      </View>
    </View>
  </View>
</View>
`;

exports[`Card with receipts attached renders correctly 1`] = `
<View>
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <View>
      <Text
        style={
          Object {
            "color": "#000000",
            "fontSize": 16,
            "fontWeight": "bold",
          }
        }
      >
        Amazon
      </Text>
      <Text
        style={
          Object {
            "color": "#000000",
            "fontSize": 14,
          }
        }
      >
        17:48
      </Text>
    </View>
    <Text
      style={
        Object {
          "color": "#000000",
          "fontSize": 16,
          "fontWeight": "500",
        }
      }
    >
      18€
    </Text>
  </View>
  <View
    style={
      Array [
        Object {
          "flexDirection": "row",
          "justifyContent": "space-between",
        },
        Object {
          "marginTop": 16,
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#E7EFFA",
            "borderRadius": 36,
            "flexDirection": "row",
            "padding": 4,
            "paddingHorizontal": 8,
          },
          undefined,
        ]
      }
    >
      <Text
        allowFontScaling={false}
        selectable={false}
        style={
          Array [
            Object {
              "color": "#000000",
              "fontSize": 16,
            },
            undefined,
            Object {
              "fontFamily": "Material Design Icons",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            Object {},
          ]
        }
      >
        󰀉
      </Text>
      <View
        style={
          Object {
            "paddingLeft": 4,
          }
        }
      />
      <View
        style={
          Object {
            "flexDirection": "row",
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#000000",
              },
              undefined,
            ]
          }
        >
          Aitor Cubeles Torres
        </Text>
      </View>
    </View>
    <View
      style={
        Object {
          "flexDirection": "row",
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "backgroundColor": "#E7EFFA",
              "borderRadius": 36,
              "flexDirection": "row",
              "padding": 4,
              "paddingHorizontal": 8,
            },
            Object {
              "marginLeft": 8,
            },
          ]
        }
      >
        <Text
          allowFontScaling={false}
          selectable={false}
          style={
            Array [
              Object {
                "color": "#000000",
                "fontSize": 16,
              },
              undefined,
              Object {
                "fontFamily": "Material Design Icons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
        >
          󰑉
        </Text>
        0
        0
      </View>
    </View>
  </View>
</View>
`;
